scipy/signal/tests/test_signaltools.py

```diff
@@ -2800,6 +2800,188 @@ def test_hilbert2_types(self, dtype):
+                          msg="Residual calculation error")
+
+        ze_1, zr_1 = envelope(z, (1, 3), residual='all', squared=False)
+        self.assert_close(ze_1**2, ze2_0,
+                          msg="Unsquared versus Squared envelope calculation error")
+        self.assert_close(zr_1, zr_0,
+                          msg="Unsquared versus Squared residual calculation error")
+
+        ze2_2, zr_2 = envelope(z, (1, 3), residual='all', squared=True, n_out=3*n)
+        self.assert_close(ze2_2[::3], ze2_0,
+                          msg="3x up-sampled envelope calculation error")
+        self.assert_close(zr_2[::3], zr_0,
+                          msg="3x up-sampled residual calculation error")
+
+        ze2_3, zr_3 = envelope(z, (1, 3), residual='lowpass', squared=True)
+        self.assert_close(ze2_3, ze2_0,
+                          msg="`residual='lowpass'` envelope calculation error")
+        self.assert_close(sp_fft.rfft(zr_3), np.array([4, 0, 0, 0, 0]).astype(complex),
+                          msg="`residual='lowpass'` residual calculation error")
+
+        ze2_4 = envelope(z, (1, 3), residual=None, squared=True)
+        self.assert_close(ze2_4, ze2_0,
+                          msg="`residual=None` envelope calculation error")
+
+        # compare complex analytic signal to real version
+        Z_a = np.copy(Z)
+        Z_a[1:] *= 2
+        z_a = sp_fft.ifft(Z_a, n=n)  # analytic signal of Z
+        self.assert_close(z_a.real, z,
+                          msg="Reference analytic signal error")
+        ze2_a, zr_a = envelope(z_a, (1, 3), residual='all', squared=True)
+        self.assert_close(ze2_a, ze2_0,
+                          msg="Complex envelope calculation error")
+        self.assert_close(sp_fft.fft(zr_a), np.array(Zr_a).astype(complex),
+                          msg="Complex residual calculation error")
+
+    @pytest.mark.parametrize(
+        "               Z,        bp_in,     Ze2_desired,      Zr_desired",
+        [([1, 0, 2, 2, 0],    (1, None), [4, 2, 0, 0, 0], [1, 0, 0, 0, 0]),
+         ([4, 0, 2, 0, 0],    (0, None), [4, 0, 2, 0, 0], [0, 0, 0, 0, 0]),
+         ([4, 0, 0, 2, 0], (None, None), [4, 0, 0, 2, 0], [0, 0, 0, 0, 0]),
+         ([0, 0, 2, 2, 0],       (1, 3), [2, 0, 0, 0, 0], [0, 0, 0, 2, 0]),
+         ([4, 0, 2, 2, 0],      (-3, 3), [4, 0, 2, 0, 0], [0, 0, 0, 2, 0]),
+         ([4, 0, 3, 4, 0],    (None, 1), [2, 0, 0, 0, 0], [0, 0, 3, 4, 0]),
+         ([4, 0, 3, 4, 0],    (None, 0), [0, 0, 0, 0, 0], [4, 0, 3, 4, 0])])
+    def test_envelope_real_signals(self, Z, bp_in, Ze2_desired, Zr_desired):
+        """Test envelope calculation with real-valued test signals.
+
+        The comparisons are performed in the Fourier space, since it makes evaluating
+        the bandpass filter behavior straightforward. Note that also the squared
+        envelope can be easily calculated by hand, if one recalls that coefficients of
```

Comment:
Not-squared envelope can also be calculated by hand. I think there should be tests covering squared=False too.

Reply:
From a testing perspective, the only difference between a squared and a non-squared envelope is [this line]:
```python
    z_env = np.abs(z_bb) if not squared else z_bb.real ** 2 + z_bb.imag ** 2
```
It is verified once in [test_envelope_verify_parameters()], which should be sufficient, IMHO.

BTW, I am not aware how to easily determine the spectrum of a square-root of a sine wave by hand. Square-rooting introduces an infinite amount of harmonics to each sinusoid.

[this line]: https://github.com/scipy/scipy/blob/2838e7d52b1b623c6c55448fdeb0f98b955d8a1a/scipy/signal/_signaltools.py#L2702
[test_envelope_verify_parameters()]: https://github.com/scipy/scipy/blob/2838e7d52b1b623c6c55448fdeb0f98b955d8a1a/scipy/signal/tests/test_signaltools.py#L2849